Sources/SwiftRefactor/ExpandEditorPlaceholder.swift

```diff
@@ -0,0 +1,347 @@
+  /// falling back to the label with the same conditions, and finally just the
+  /// placeholder-ed type otherwise.
+  fileprivate func expansionNameToken() -> TokenSyntax {
+    if let secondName = secondName, secondName.tokenKind != .wildcard {
+      return secondName
+    }
+
+    if let name = name, name.tokenKind != .wildcard {
+      return name
+    }
+
+    return .identifier(ExpandEditorPlaceholder.wrapInPlaceholder(type.description))
+  }
+}
+
+extension FunctionCallExprSyntax {
+  /// If the given argument is one of the last arguments that are all
+  /// function-typed placeholders and this call doesn't have a trailing
+  /// closure, then return a replacement of this call with one that uses
+  /// closures based on the function types provided by each editor placeholder.
+  /// Otherwise return nil.
+  fileprivate func expandTrailingClosurePlaceholders(ifIncluded: TupleExprElementSyntax) -> (expr: FunctionCallExprSyntax, numClosures: Int)? {
+    var includedArg = false
+    var argsToExpand = 0
+    for arg in argumentList.reversed() {
+      guard let expr = arg.expression.as(EditorPlaceholderExprSyntax.self),
+        let data = EditorPlaceholderData(token: expr.identifier),
+        case let .typed(_, type) = data,
+        type.is(FunctionTypeSyntax.self)
+      else {
+        break
+      }
+      if arg == ifIncluded {
+        includedArg = true
+      }
+      argsToExpand += 1
+    }
+
+    guard includedArg else {
+      return nil
+    }
+
+    var expandedArgs = [TupleExprElementSyntax]()
+    for arg in argumentList.suffix(argsToExpand) {
+      guard let expr = ExpandEditorPlaceholder.refactor(syntax: arg.expression.cast(EditorPlaceholderExprSyntax.self).identifier) else {
+        return nil
+      }
+
+      expandedArgs.append(
+        TupleExprElementSyntax(
+          label: arg.label ?? .wildcardToken(),
+          colon: .colonToken(),
+          expression: expr,
+          trailingComma: nil
+        )
+      )
+    }
+
+    let originalArgs = argumentList.dropLast(argsToExpand)
+    return (with(\.argumentList, TupleExprElementListSyntax(originalArgs + expandedArgs)), expandedArgs.count)
+  }
+}
+
+/// Placeholder text must start with '<#' and end with
+/// '#>'. Placeholders can be one of the following formats:
+/// ```
+///   'T##' display-string '##' type-string ('##' type-for-expansion-string)?
+///   'T##' display-and-type-string
+///   display-string
+/// ```
+///
+/// NOTE: It is required that '##' is not a valid substring of display-string
+/// or type-string. If this ends up not the case for some reason, we can consider
+/// adding escaping for '##'.
+fileprivate enum EditorPlaceholderData {
+  case basic(text: Substring)
+  case typed(text: Substring, type: TypeSyntax)
+
+  init?(token: TokenSyntax) {
+    guard ExpandEditorPlaceholder.isPlaceholder(token.text) else {
+      return nil
+    }
+
+    var text = token.text.dropFirst(2).dropLast(2)
+
+    if !text.hasPrefix("T##") {
```

Comment:
Is `T##` really the thing we want to be looking for here? I thought that the first type is the one that will be used during type checking and it can be any type. But I might be completely wrong.

Reply:
`T## is mandatory for both typechecking and expansion. Without `T##` typechecking treats it as `Void`. As an example, `foo(<#(Int) -> (Int)#>)` is not expanded to a closure (or trailing closure).

Reply:
Interesting. You learn something new every day.